Issue768 extlib upgrade

sarahjcotton opened the pull request:
Fix for issue #768

Some re-factoring was required in order to make the plugin compatible with simplesamlphp 2.0 as per the docs: https://simplesamlphp.org/docs/2.0/simplesamlphp-upgrade-notes-2.0.html

In addition, a new BINDING_HTTP_ARTIFACT const seems to cause artifact to always take priority instead of redirect. Removal rather than re-order might be preferable here but I've left it in as there are settings in the plugin that refer to artifact bindings, although I'm not sure sure how to configure and test if the two are related. I couldn't see an obvious way to achieve this programmatically, unfortunately.

A number of unit tests are skipped when I run locally due to Redis and the Prophecy library not being available, again, not sure how to test.

Otherwise, @Peterburnett has kindly offered to peer review this. Happy to make any further changes as per your guidance.

hdagheda committed:
WR #333356: Manually apply below patch due to dir change:

extlib-simplesaml: backport mods to extlib. 30632972e780ff5d8ccafb2c379864f9138d5390
Added support for IdP initiation flow #105 6ebf501a378eec10f5b268650e7399b906aeb63c
Fixed bug with Idp initiatied entityid ca343939b8c12951abcd533704ded45068a2ac99

Edited 04/10/23 for peer review of pull #774

danmarsden committed:
Remove phpmailer from simplesaml lib, see issue #469

daledavies committed:
Issue #768 - Update simplesamlphp to version 2.0.5

daledavies committed:
Update permissions.

sarahjcotton committed:
Issue #768: Remove unnecessary param being sent to sendSAML2AuthnRequest()

sarahjcotton committed:
Issue #768: Changes required after upgrade to SimpleSaml 2.0

https://simplesamlphp.org/docs/2.0/simplesamlphp-upgrade-notes-2.0.html

sarahjcotton committed:
Issue #768: Force de-prioritisation of BINDING_HTTP_ARTIFACT

sarahjcotton committed:
Issue #768: Remove support for SAML 1.0/1.1 protocols.

sarahjcotton committed:
Issue #768: Fix unit test

sarahjcotton committed:
Issue #768: Version bump.

Peterburnett commented on .extlib/simplesamlphp/modules/saml/src/Auth/Source/SP.php:
```diff
@@ -306,6 +306,22 @@ public function getHostedMetadata(): array
      */
     public function getIdPMetadata(string $entityId): Configuration
     {
+        assert(is_string($entityId));
```
This assertion should likely be removed, the typeguard will force it to a string either way. Leftovers from debugging?

Peterburnett commented on .extlib/simplesamlphp/modules/saml/src/Auth/Source/SP.php:
```diff
@@ -306,6 +306,22 @@ public function getHostedMetadata(): array
     {
+        assert(is_string($entityId));
+
+        global $saml2auth;
```
Can we get a modification comment highlighting this as a auth_saml2 modification

Peterburnett commented on .extlib/simplesamlphp/modules/saml/src/Auth/Source/SP.php:
```diff
@@ -629,16 +648,22 @@ private function startSSO2(Configuration $idpMetadata, array $state): void
-                    Constants::BINDING_HTTP_ARTIFACT,
                     Constants::BINDING_HTTP_REDIRECT,
                     Constants::BINDING_HTTP_POST,
+                    Constants::BINDING_HTTP_ARTIFACT,
```
Additionally, can we get a modification comment above this block mentioning the reordering to put artifact last to maintain existing behaviour.

Peterburnett commented on .extlib/simplesamlphp/modules/saml/src/Auth/Source/SP.php:
```diff
@@ -1118,9 +1143,19 @@ public function handleLogout(string $idpEntityId): void
     public static function handleUnsolicitedAuth(string $authId, array $state, string $redirectTo): void
     {
+        global $SESSION, $saml2auth;
+        assert(is_string($authId));
```
Remove these assertions, type guards on the method will enforce this

Peterburnett commented on .extlib/simplesamlphp/src/SimpleSAML/Utils/HTTP.php:
```diff
@@ -951,7 +951,18 @@ public function parseQueryString(string $query_string): array
      */
     public function redirectTrustedURL(string $url, array $parameters = []): void
     {
+        if (!is_string($url) || !is_array($parameters)) {
```
Is this required? The parameters are typed, this should never be possible now

Peterburnett commented on classes/store.php:
```diff
@@ -88,7 +88,7 @@ public function get($type, $key) {
+    public function set(string $type, string $key, $value, ?int $expire = null): void {
         global $DB;
 
         assert(is_string($type));
```
I dont think this assertion is required at all now that we are type guarded in the parameters

sarahjcotton replied:
I kept in one of the assertions but changed it slightly as it's checking for a particular value not covered by the param types.

Peterburnett commented on classes/store.php:
```diff
@@ -133,7 +133,7 @@ public function set($type, $key, $value, $expire = null) {
+    public function delete(string $type, string $key): void {
         global $DB;
 
         assert(is_string($type));
```
Same for these assertions

Peterburnett commented on db/upgrade.php:
```diff
@@ -388,5 +388,25 @@ function xmldb_auth_saml2_upgrade($oldversion) {
+        $a = array_search('urn:oasis:names:tc:SAML:1.0:profiles:browser-post', $protocols);
+        $b = array_search('urn:oasis:names:tc:SAML:1.0:profiles:artifact-01', $protocols);
+
+        if ($a) {
```
Can this get a quick explainer comment for anyone digging into the upgrade behaviours

Peterburnett commented on sp/saml2-acs.php:
```diff
@@ -31,7 +31,12 @@
+    $session = \SimpleSAML\Session::getSessionFromRequest();
+    $controller = new \SimpleSAML\Module\saml\Controller\ServiceProvider($config, $session);
+    $acs = $controller->assertionConsumerService($saml2auth->spname);
+    $runnable = new \SimpleSAML\HTTP\RunnableResponse($acs->getCallable(), $acs->getArguments());
```
do we need this $runnable line? not sure if it needs to be wired into anything

Peterburnett commented on sp/saml2-logout.php:
```diff
@@ -55,8 +55,12 @@
+    $session = \SimpleSAML\Session::getSessionFromRequest();
+    $controller = new \SimpleSAML\Module\saml\Controller\ServiceProvider($config, $session);
+    $acs = $controller->singleLogoutService($saml2auth->spname);
+    $runnable = new \SimpleSAML\HTTP\RunnableResponse($acs->getCallable(), $acs->getArguments());
```
same as above, do we need to do anything with this $runnable var?

Peterburnett commented on db/upgrade.php:
```diff
@@ -388,5 +388,25 @@ function xmldb_auth_saml2_upgrade($oldversion) {
+        }
+
+        $protocols = implode(',', $protocols);
+        set_config('assertionsconsumerservices', $protocols,'auth_saml2');
```
Need a space here after $protocols, causing a CI fail

Peterburnett commented:
Hi @sarahjcotton 

All is largely looking good. Couple of comments could help to explain some stuff, and I think that there are a bunch of asserts in place that we dont need anymore as they are superseded by type guards on the parameters. Additionally there is a small syntax issue causing CI fails that ive put a comment on.

Other than that, I've seen this working with keycloak, so I think this is good to start rolling out. Anything else will have to be found in whatever complex integrations surface the issues.

sarahjcotton commented:
Thanks so much Peter; I've addressed all the issues you flagged if you wouldn't mind reviewing those,  however whilst trying to run some unit tests yesterday I set the $CFG vars up for Redis and on trying to login using SAML today a bunch of new errors have been thrown that need fixing, so marking this PR as draft for now as it's not ready to be merged.

sarahjcotton committed:
Issue #768: Fix redis_store class to use factory StoreInterface

sarahjcotton commented:
New commit added to fix the redis_store class. I then also managed to get the redis unit tests to run locally and they all passed:

```
./vendor/bin/phpunit --verbose --filter auth_saml2_redis_store_testcase
Runtime:       PHP 8.0.29
Configuration: /var/www/moodle41/phpunit.xml

.......                                                             7 / 7 (100%)

Time: 00:01.598, Memory: 324.00 MB

OK (7 tests, 7 assertions)

```

Peterburnett commented:
Hi @sarahjcotton 

This is looking much better now, things are looking really clean. The CI has exposed something interesting, in that PHP 7.4 is now the minimum PHP version supported. All 7 versions are EOL, so I dont think its a major issue to set the minimum to 7.4, but we should update the README table to reflect that, and additionally update the CI grid so that PHP 7.2 and 7.3 are excluded (see https://github.com/catalyst/catalyst-moodle-workflows#with-options)

I think thats the last thing before we can land this one.

danmarsden commented:
Great work Sarah!!

+1 to merge after 
1) update readme to state that php 7.4 and higher is required on the 3.9 stable branch as Peter mentioned.
2) add the min_php setting to the github actions workflow config to only run 7.4 and higher tests (see: https://github.com/catalyst/catalyst-moodle-workflows/#with-options)

sarahjcotton committed:
Issue #768: Fix ci php version support

sarahjcotton commented:
Thanks both, hopefully that's good to go now!

cameron1729 commented on locallib.php:
```diff
@@ -57,16 +57,16 @@ function auth_saml2_get_sp_metadata($baseurl = '') {
     $entityId = $source->getEntityId();
     $spconfig = $source->getMetadata();
-    $store = SimpleSAML\Store::getInstance();
+    \SimpleSAML\Store\StoreFactory::getInstance('\\auth_saml2\\store');
```
A bit confused by this. It doesn't seem that `$store` was ever used. So could we not just delete this line?
